✨  BT-1418 integrate foundry & hardhat + migrating solidity tests

mohamadhammoud opened the pull request.

mohamadhammoud committed:
✨ integrate foundry alongside hardhat for dual testing framework

- Add foundry.toml configuration
- Reorganize test structure: test/forge/ and test/hardhat/
- Add Foundry scripts to package.json (build:forge, test:forge, coverage:forge)
- Update hardhat.config.ts with evmVersion for compatibility
- Add GitHub workflows and coverage scripts from template
- Add Husky hooks (pre-commit, post-checkout, post-merge)
- Add .vscode/settings.json for better IDE support
- Disable coverage checks until tests are migrated (commented with TODO)

This migration is based on TokenySolutions/contracts-template and enables
gradual migration of tests from Hardhat (TypeScript) to Foundry (Solidity).

Related to: BT-1418

mohamadhammoud committed:
✅ trex factory tests

mohamadhammoud committed:
⚰️ remove trex suit

mohamadhammoud committed:
feat(forge): reorganize Foundry test structure and sync GitHub workflows

- Move TREXFactorySetup to helpers folder
- Rename TREXSuite.t.sol to TREXFactory.t.sol
- Update fixture to inherit from Test for proper Foundry pattern
- Transfer ownership to deployer address to match Hardhat behavior
- Sync GitHub workflows with package.json scripts
- Update test.yml to use npm run build and npm run test
- Update push_checking.yml to use unified npm run lint
- Update forge_coverage.sh to use npm run coverage

mohamadhammoud committed:
fix(ci): install Foundry and skip husky in push_checking workflow

- Add Foundry installation step before npm ci
- Set HUSKY=0 to skip husky setup in CI environment
- Fixes 'forge: not found' error during npm prepare script

pgonday commented on test/forge/helpers/TREXFactorySetup.sol:
```diff
@@ -0,0 +1,104 @@
+    /// Creates a reference Implementation Authority (isReference = true)
+    function setUp() public {
+        // Create standard test addresses
+        deployer = makeAddr("deployer");
```
could be in constructor instead of setup, or even directly at variable declaration 
`address public deployer = makeAddr("deployer");
`

mohamadhammoud replied:
[bd99a26](https://github.com/ERC-3643/ERC-3643/pull/76/commits/bd99a26bf6940c241df4f93ca010c2baa603e181)

pgonday commented on test/forge/helpers/TREXFactorySetup.sol:
```diff
@@ -0,0 +1,104 @@
+            TREXFactoryHelper.deploy(implementationAuthoritySetup.implementationAuthority, onchainidSetup.idFactory);
+
+        // transfer ownership to deployer after setting up everything
+        Ownable(address(implementationAuthoritySetup.implementationAuthority)).transferOwnership(deployerAddress);
```
Not sure if we just transfer ownership or if we directly create the contracts with deployer ?

pgonday commented on test/forge/TREXFactory.t.sol:
```diff
@@ -0,0 +1,34 @@
+        assertNotEq(address(getIdFactory()), address(0), "IdFactory should be deployed");
+    }
+
+    function test_StandardAddressesSet() public view {
```
This is testing test stuff, not Factory test

mohamadhammoud replied:
[2feeb33](https://github.com/ERC-3643/ERC-3643/pull/76/commits/2feeb3371578ce83d5cca6ddcf0efbe3397ef29e)

pgonday commented on test/forge/helpers/IdentityFactoryHelper.sol:
```diff
@@ -0,0 +1,33 @@
+import { IdFactory } from "@onchain-id/solidity/contracts/factory/IdFactory.sol";
+import { ImplementationAuthority } from "@onchain-id/solidity/contracts/proxy/ImplementationAuthority.sol";
+
+/// @title IdentityFactoryHelper
```
Useless natspec in tests

mohamadhammoud replied:
[2f7d96e](https://github.com/ERC-3643/ERC-3643/pull/76/commits/2f7d96e38c9e3a1a5c0c387b32395d54bc7153a4)

mohamadhammoud committed:
🔧 keep hardhat settings

mohamadhammoud committed:
✅ create addressess directly

mohamadhammoud committed:
✅ removed test_StandardAddressesSet

mohamadhammoud committed:
✅ fix lint

mohamadhammoud committed:
💡 removed title natspecs

mohamadhammoud committed:
✅ migrated trex factort tests

mohamadhammoud committed:
✅ added test module multicall

mohamadhammoud committed:
✅ migrate agent role test

mohamadhammoud committed:
✅ migrate claim topics registry

mohamadhammoud committed:
✅ migrated trusted issuer registry

mohamadhammoud committed:
✅ migrate identity registry storage

mohamadhammoud committed:
✅ migrated identity registry tests

mohamadhammoud committed:
✅ migrated implementation authority test

mohamadhammoud committed:
✅ migrated token agen restrictions

mohamadhammoud committed:
✅ migrated token information test

mohamadhammoud committed:
✅ migrated token recover

mohamadhammoud committed:
✅ migrated token permit

mohamadhammoud committed:
✅ migrated token transfer

mohamadhammoud committed:
✅ migrated utilities upgrade

mohamadhammoud committed:
✅ migrated freeze check

mohamadhammoud committed:
✅ migrated compliance check in utilities test

mohamadhammoud committed:
✅ migrated eligibility check

mohamadhammoud committed:
✅ migrated transfer check in utilities

mohamadhammoud committed:
✅ migrated compliance test

pgonday commented on test/forge/compliance/Compliance.t.sol:
```diff
@@ -0,0 +1,850 @@
+import { TREXFactorySetup } from "test/forge/helpers/TREXFactorySetup.sol";
+
+// Error for InvalidInitialization (OpenZeppelin v5 doesn't export this)
+error InvalidInitialization();
```
Import `Initializable` and use `Initializable.InvalidInitialization`

pgonday commented on test/forge/compliance/Compliance.t.sol:
```diff
@@ -0,0 +1,850 @@
+contract ComplianceTest is TREXFactorySetup {
+
+    // Event declarations for expectEmit
+    event ComplianceAdded(address indexed _compliance);
```
Why are you redefining events instead of using the ones available in interface contracts ?

mohamadhammoud replied:
[f37007b](https://github.com/ERC-3643/ERC-3643/pull/76/commits/f37007b27de9b31f115138b9fe58ac34e52a47b6)

pgonday commented on test/forge/compliance/Compliance.t.sol:
```diff
@@ -0,0 +1,850 @@
+        // Register alice and bob identities and mint tokens (for tests that need them)
+        ImplementationAuthority identityImplementationAuthority =
+            ImplementationAuthority(onchainidSetup.idFactory.implementationAuthority());
+        IIdentity aliceIdentity = IIdentity(address(new IdentityProxy(address(identityImplementationAuthority), alice)));
```
Maybe use IdFactory.createIdentity()

mohamadhammoud replied:
[0547ef2](https://github.com/ERC-3643/ERC-3643/pull/76/commits/0547ef29fa0a27059c7ab3ce7cfb00f4f6a04c29)

pgonday commented on test/forge/compliance/Compliance.t.sol:
```diff
@@ -0,0 +1,850 @@
+    function test_destroyed_Success_WhenAmountGreaterThanZero() public {
+        _setupComplianceBoundToWallet();
+
+        vm.prank(charlieWallet);
```
`destroyed` is `onlyToken`, the test will be more understandable with a `token` instead of `charlieWallet`

mohamadhammoud replied:
[0d5b1bf](https://github.com/ERC-3643/ERC-3643/pull/76/commits/0d5b1bf8bdb075fad05e1d11f66a5d930189785c)

mohamadhammoud committed:
✅ migrated gateway tests

pgonday commented on test/forge/token/TokenPermit.t.sol:
```diff
@@ -0,0 +1,179 @@
+        super.setUp();
+
+        // Deploy token suite
+        ITREXFactory.TokenDetails memory tokenDetails = ITREXFactory.TokenDetails({
```
This code is duplicated on all token tests, either add it to the base test, or create a token base test

mohamadhammoud replied:
[35b3bd7](https://github.com/ERC-3643/ERC-3643/pull/76/commits/35b3bd72a4b8c6a24c573e727840290fe25035ae)

pgonday commented on test/token/TokenRecovery.t.sol:
```diff
@@ -0,0 +1,262 @@
+    // Token suite deployed in setUp()
+    address public tokenAddress;
+    Token public token;
+    IdentityRegistry public identityRegistry;
```
Use only token, and cast to address when needed

mohamadhammoud replied:
[35b3bd7](https://github.com/ERC-3643/ERC-3643/pull/76/commits/35b3bd72a4b8c6a24c573e727840290fe25035ae)

mohamadhammoud committed:
🔥 removed already defined events

mohamadhammoud committed:
🗑️ imported Initializable error from openzeppelin

mohamadhammoud committed:
➕ added already defined events and errors

mohamadhammoud committed:
✅ added base test token

mohamadhammoud committed:
✅ used idFactory

mohamadhammoud committed:
♻️ refactored compliance test

pgonday commented on test/forge/authorities/TREXImplementationAuthority.t.sol:
```diff
@@ -0,0 +1,528 @@
+        ImplementationAuthorityHelper.ImplementationAuthoritySetup memory otherIASetup =
+            ImplementationAuthorityHelper.deploy(true);
+
+        vm.prank(address(this));
```
Remove all these useless `vm.prank(address(this))` (x40)

mohamadhammoud replied:
[5b280da](https://github.com/ERC-3643/ERC-3643/pull/76/commits/5b280da4568b93c4b9ec0ad001c435c22c71aa8d)

pgonday commented on test/authorities/TREXImplementationAuthority.t.sol:
```diff
@@ -0,0 +1,528 @@
+        // Deploy a new factory using this IA
+        TREXFactory newFactory = new TREXFactory(address(getTREXImplementationAuthority()), address(getIdFactory()));
+        vm.prank(address(this));
+        Ownable(address(newFactory)).transferOwnership(deployer);
```
unnecessary cast

mohamadhammoud replied:
[5b280da](https://github.com/ERC-3643/ERC-3643/pull/76/commits/5b280da4568b93c4b9ec0ad001c435c22c71aa8d)

pgonday commented on test/authorities/TREXImplementationAuthority.t.sol:
```diff
@@ -0,0 +1,528 @@
+    function test_fetchVersion_RevertWhen_AlreadyFetched() public {
+        TREXFactory factory = new TREXFactory(address(getTREXImplementationAuthority()), address(getIdFactory()));
+        vm.prank(address(this));
+        Ownable(address(factory)).transferOwnership(deployer);
```
There is a lot of unnecessary casts, also for `otherIA`

mohamadhammoud replied:
[5b280da](https://github.com/ERC-3643/ERC-3643/pull/76/commits/5b280da4568b93c4b9ec0ad001c435c22c71aa8d)

mohamadhammoud committed:
⚰️ remoced useless vm.prank

mohamadhammoud committed:
🔧 deleted hardhat configurations

mohamadhammoud committed:
💚 fixed ci

mohamadhammoud committed:
refactor: remove 100% coverage requirement from forge_coverage.sh

- Remove strict 100% coverage checks that were causing workflow failures
- Script now only generates and displays coverage report without failing
- Aligns with previous behavior where coverage was informational only

mohamadhammoud committed:
✅ resolve getVerifiedDetails

mohamadhammoud committed:
✅ applied _getEligibility and remove via-ir from foundry.toml

mohamadhammoud committed:
✅ cover tokena and token permit

mohamadhammoud committed:
✅ cover compliance

mohamadhammoud committed:
✅ cover identity registry

mohamadhammoud committed:
✅ increased trex gateway coverage

mohamadhammoud committed:
✅ increase coverage for trex factory

mohamadhammoud committed:
✅ increased AbstractModuleUpgradeable.sol coverage

mohamadhammoud committed:
⚰️ removed abstract module

mohamadhammoud committed:
🔧 updted generate coverage

mohamadhammoud committed:
✅ coverage 100% identity registry

mohamadhammoud committed:
✅ coverage 100% trex gate way

mohamadhammoud committed:
✅ coverage trex factory

mohamadhammoud committed:
✅ coverage 100% AbstractModuleUpgradeable by test module

mohamadhammoud committed:
✅ coverage IAFactory.sol

mohamadhammoud committed:
✅ coverage 100% for implementation authority

mohamadhammoud committed:
✅ coverage 1005 abstract proxy

mohamadhammoud committed:
✅ coverage 100% claim topic registry

mohamadhammoud committed:
✅ coverage 100% identity registry storage

mohamadhammoud committed:
✅ coverage 100% modular compliance proxy

mohamadhammoud committed:
✅ coverage trusted issuer registry proxy

mohamadhammoud committed:
✅ coverage 100% identity registry

mohamadhammoud committed:
✅ coverage 100% token proxy

mohamadhammoud committed:
🚨 resolve linter in token information

mohamadhammoud committed:
✅ coverage 100% roles

mohamadhammoud committed:
🚨 resolve linter

mohamadhammoud committed:
🔧 added no coverage for test, _testContracts and  utils

mohamadhammoud committed:
✅ coverage 100% for token.sol

mohamadhammoud committed:
✅ coverage 100% for modular compliance

mohamadhammoud committed:
✅ coverage 100% trusted issuer registry

pgonday reviewed (changes requested):
- InterfaceIdCalculator and UtilityChecker are missing some coverage (94%, 92%)
- The `package.json` should not have been removed, it contains all the commands in the scripts section, including the prepare command with husky & forge init.

mohamadhammoud committed:
🔧 restore package.json

mohamadhammoud committed:
✅ coverage 100% for utlity checker

mohamadhammoud committed:
🚨 fix linter
